File: data/readable_format/fortran/file1620.f90
PROGRAM TestILADLR
  IMPLICIT NONE
  INTEGER :: M, N, LDA, result
  DOUBLE PRECISION, ALLOCATABLE :: A(:,:)

  INTERFACE
    INTEGER FUNCTION ILADLR(M, N, A, LDA)
      INTEGER, INTENT(IN) :: M, N, LDA
      DOUBLE PRECISION, INTENT(IN) :: A(LDA, *)
    END FUNCTION ILADLR
  END INTERFACE

  M = 4
  N = 3
  LDA = 4
  ALLOCATE(A(LDA,N))

  ! Initialize matrix A
  A = 0.0D+0
  A(1,1) = 1.0D+0
  A(2,3) = 2.0D+0
  A(3,2) = 3.0D+0
  ! The last row is all zeros

  result = ILADLR(M, N, A, LDA)
  PRINT *, "The last non-zero row is:", result

  DEALLOCATE(A)
END PROGRAM TestILADLR

! The ILADLR function definition
INTEGER FUNCTION ILADLR(M, N, A, LDA)
  INTEGER, INTENT(IN) :: M, N, LDA
  DOUBLE PRECISION, INTENT(IN) :: A(LDA, *)
  DOUBLE PRECISION :: ZERO
  PARAMETER (ZERO = 0.0D+0)
  INTEGER :: I, J

  IF (M == 0) THEN
     ILADLR = M
  ELSE IF (A(M, 1) /= ZERO .OR. A(M, N) /= ZERO) THEN
     ILADLR = M
  ELSE
     ILADLR = 0
     DO J = 1, N
        I = M
        DO WHILE ((A(MAX(I,1), J) == ZERO) .AND. (I >= 1))
           I = I - 1
        END DO
        ILADLR = MAX(ILADLR, I)
     END DO
  END IF
  RETURN
END FUNCTION ILADLR

File: data/readable_format/cpp/file1620.cpp
#include <iostream>
#include <vector>
#include <algorithm> // For std::max

// Declaration and definition of the ILADLR function
int ILADLR(int M, int N, double** A, int LDA) {
    const double ZERO = 0.0;

    if (M == 0) {
        return M;
    } else if (A[M-1][0] != ZERO || A[M-1][N-1] != ZERO) {
        return M;
    } else {
        int result = 0;
        for (int J = 0; J < N; ++J) {
            int I = M;
            while ((A[std::max(I-1, 0)][J] == ZERO) && (I >= 1)) {
                --I;
            }
            result = std::max(result, I);
        }
        return result;
    }
}

// Test program
int main() {
    int M = 4, N = 3, LDA = 4;
    std::vector<std::vector<double>> matrix(M, std::vector<double>(N, 0.0));

    // Initialize matrix
    matrix[0][0] = 1.0;
    matrix[1][2] = 2.0;
    matrix[2][1] = 3.0;
    // The last row is all zeros

    // Convert to double pointer for ILADLR function
    double** A = new double*[LDA];
    for (int i = 0; i < LDA; ++i) {
        A[i] = matrix[i].data();
    }

    int result = ILADLR(M, N, A, LDA);
    std::cout << "The last non-zero row is: " << result << std::endl;

    // Clean up
    delete[] A;

    return 0;
}